File: 32/src/macros.c
/* CHOICE macro
 * Choice(e, f, g) := (e & f) ^ (~e & g)
 * Returns bits from f when e=1, bits from g when e=0
 */
#define CHOICE(e, f, g) (((e) & (f)) ^ (~(e) & (g)))

/* MEDIAN macro
 * Median(e, f, g) := (e & f) ^ (e & g) ^ (f & g)
 * Majority of bits among e, f, g
 */
#define MEDIAN(e, f, g) (((e) & (f)) ^ ((e) & (g)) ^ ((f) & (g)))

/* ROTATE macro
 * Right rotate x by n bits (assuming 64-bit values)
 */
#define ROTATE(x, n) (((x) >> (n)) | ((x) << (64 - (n))))

File: 32/src/macros.rs
#[macro_export]
macro_rules! choice {
    ($x:expr, $y:expr, $z:expr) => {
        ($x & $y) ^ ((!$x) & $z)
    };
}

#[macro_export]
macro_rules! median {
    ($x:expr, $y:expr, $z:expr) => {
        ($x & $y) | ($x & $z) | ($y & $z)
    };
}

#[macro_export]
macro_rules! rotate {
    ($x:expr, $n:expr) => {
          (($x) >> (($n) % 64))
    };
}
